lpass.1.txt
:man source:   lpass
:man manual:   lpass

LPASS(1)
========


NAME
----
lpass - command line interface for LastPass

SYNOPSIS
--------
[verse]
*lpass* [ --version, -v | --help, -h ]
*lpass* <subcommand> [<args>]

DESCRIPTION
-----------
'lpass' is a simple command line interface to LastPass. It is comprised of
several subcommands:

[verse]
 lpass *login* [--trust] [--plaintext-key [--force, -f]] [--color=auto|never|always] USERNAME
 lpass *logout* [--force, -f] [--color=auto|never|always]
 lpass *passwd*
 lpass *show* [--sync=auto|now|no] [--clip, -c] [--quiet, -q] [--expand-multi, -x] [--json, -j] [--all|--username|--password|--url|--notes|--field=FIELD|--id|--name|--attach=ATTACHID] [--basic-regexp, -G|--fixed-strings, -F] [--color=auto|never|always] {NAME|UNIQUEID}*
 lpass *ls* [--sync=auto|now|no] [--long, -l] [-m] [-u] [--color=auto|never|always] [GROUP]
 lpass *mv* [--sync=auto|now|no] [--color=auto|never|always] {UNIQUENAME|UNIQUEID} GROUP
 lpass *add* [--sync=auto|now|no] [--non-interactive] {--name|--username, -u|--password, -p|--url|--notes|--field=FIELD|--note-type=NOTETYPE} [--color=auto|never|always] {NAME|UNIQUEID}
 lpass *edit* [--sync=auto|now|no] [--non-interactive] {--name|--username, -u|--password, -p|--url|--notes|--field=FIELD} [--color=auto|never|always] {NAME|UNIQUEID}
 lpass *generate* [--sync=auto|now|no] [--clip, -c] [--username=USERNAME] [--url=URL] [--no-symbols] [--color=auto|never|always] {NAME|UNIQUEID} LENGTH
 lpass *duplicate* [--sync=auto|now|no] [--color=auto|never|always] {UNIQUENAME|UNIQUEID}
 lpass *rm* [--sync=auto|now|no] [--color=auto|never|always] {UNIQUENAME|UNIQUEID}
 lpass *status* [--quiet, -q] [--color=auto|never|always]
 lpass *sync* [--background, -b] [--color=auto|never|always]
 lpass *import* [--sync=auto|now|no] [--keep-dupes] [FILENAME]
 lpass *export* [--sync=auto|now|no] [--color=auto|never|always] [--fields=FIELDLIST]
 lpass *share* *userls* SHARE
 lpass *share* *useradd* [--read-only=[true|false]] [--hidden=[true|false]] [--admin=[true|false]] SHARE USERNAME
 lpass *share* *usermod* [--read-only=[true|false]] [--hidden=[true|false]] [--admin=[true|false]] SHARE USERNAME
 lpass *share* *userdel* SHARE USERNAME
 lpass *share* *create* SHARE
 lpass *share* *rm* SHARE
 lpass *share* *limit* [--deny|--allow] [--add|--rm|--clear] SHARE USERNAME [sites]

Synchronization
~~~~~~~~~~~~~~~
The '--sync' options control when the current operation involves a
synchronization with the server. If 'now' is set, and the command makes a
change, the change is synchronized before the command exits. If 'now' is set,
and the command displays a value, the local cache is synchronized before the
value is shown. If 'now' is set, and the command is otherwise successful, but
synchronization fails, the command will return an error. If 'auto' is set,
and the command makes a change, the change is synchronized to the server in
the background. If 'auto' is set, and the command displays a value, the local
cache is synchronized before the value is shown only if the local cache is
more than 5 seconds (or 'LPASS_AUTO_SYNC_TIME' seconds, if set) old. If 'no'
is set, the command will not interact with the server, unless there is a
current upload queue being processed. Any local changes that are not
synchronized with the server will exist in a queue of timestamped requests
which will be synchronized on the next occurring synchronization.

The 'sync' command forces a synchronization of the local cache with the
LastPass servers, and does not exit until the local cache is synchronized or
until an error occurs. Alternatively, if '--background' is specified, the
synchronization occurs in a daemonized process.

Agent
~~~~~
An agent process will be spawned in the background on a first successful
command, and all subsequent commands will use the agent for decryption,
instead of asking a user for a password. The agent will quit after one hour,
unless the 'LPASS_AGENT_TIMEOUT' environment variable is set to an alternative
number of seconds in which to quit, or 0 to never quit. If the environment
variable 'LPASS_AGENT_DISABLE' is set to 1, the agent will not be used.

Password Entry
~~~~~~~~~~~~~~
The *pinentry* program, part of *gpg2*(1), may be used for inputting
passwords if it is installed. A custom path to the *pinentry* program can be
provided by the 'LPASS_PINENTRY' environment variable.

If *pinentry* program is unavailable, or if the 'LPASS_DISABLE_PINENTRY'
environment variable is set to 1, passwords will be read from standard input and a
prompt will be displayed on standard error.

The program used for inputting passwords may also be configured by setting the
'LPASS_ASKPASS' environment variable. 'LPASS_ASKPASS' is expected to be a
binary that produces a prompt using its first command-line argument, and
outputs the entered password to standard out.  ssh-askpass implements this
protocol, as does the following shell script:

[verse]
 #!/bin/bash
 echo -n "$*: " >/dev/stderr
 stty -echo
 read answer
 stty echo
 echo $answer

Entry Specification
~~~~~~~~~~~~~~~~~~~
Commands that take a 'UNIQUENAME' will fail if the provided name is used
multiple times, and return an error. Commands may alternatively take a
'UNIQUEID', which will be the integer 'ID' provided by LastPass for identifying
entries uniquely. Commands that take either a 'NAME' or a 'UNIQUEID' will create a
new entry if a 'NAME' is specified and otherwise overwrite an existing entry if
'UNIQUEID' is specified.

Logging In
~~~~~~~~~~
The 'login' subcommand will initialize a local cache and configuration folder,
then attempt to authenticate itself with the LastPass servers, using the
provided command line credentials or by interactively prompting (in the case of
multifactor or an unprovided password). The '--trust' option will cause
subsequent logins to not require multifactor authentication. If the
'--plaintext-key' option is specified, the decryption key will be saved to the
hard disk in plaintext.  Please note that use of this option is discouraged
except in limited situations, as it greatly decreases the security of data.

The 'logout' subcommand will remove the local cache and stored encryption
keys. It will prompt the user to confirm, unless '--force' is specified.

The 'passwd' subcommand may be used to change your LastPass password:
it will prompt for the old and new password and then re-encrypt all records
with the newly derived key.

Viewing
~~~~~~~
The 'show' subcommand will display a password or selected field.

By default, the site you specify with the 'show' subcommand must
*exactly* match the name of the site.  If the '--fixed-strings' or '-F'
option is set, then the 'show' subcommand will find a site containing
that *exact* substring; if the '--basic-regexp' or '-G' option is set,
then the 'show' subcommand will find a site matching a case-insensitive
regular expression.

By default if your 'show' subcommand matches more than one site, then
the 'show' subcommand will generate a warning and display the names of
matching sites but no other information.  The '--expand-multi' or '-x'
option will instead show the requested information from all of the
matching sites.

The 'ls' subcommand will list names in groups in a tree structure. If
the '--long' or '-l' option is set, then also list the last modification
time.  The '-u' option may be passed to show the last use (last touch) time
instead, if available. Both times are in GMT.

Passing '--json' to 'show' will generate json output instead of
human-readable text.

In addition to using the built-in formats, both 'show' and 'ls' subcommands
support printf-style format strings by using the '--format' option with
the following placeholders:

* %ai: account id
* %an: account name
* %aN: account name including path
* %au: account user
* %ap: account password
* %am: account modification time
* %aU: account last touch time
* %as: account share name
* %ag: account group name
* %al: account URL
* %fn: field name (for 'show')
* %fv: field value (for 'show')

A slash can be added between the '%' and the placeholder to indicate that a
slash should be appended, only if the printed value is expanded to a non-empty
string.  For example, this command will properly show the full path to
an account: `lpass ls --format="%/as%/ag%an"`. In the same way, an underscore may
be added between the '%' and the placeholder to indicate that the value should
be made file-system-safe by replacing forward slashes, tabs, carriage returns,
and new lines with an underscore. This may be combined with the slash. For 
example: `lpass show --format="%/_as%/_ag%_an"`.

Modifying
~~~~~~~~~
The 'edit' subcommand will edit the selected field. If '--non-interactive' is not
set, the selected field will be edited using 'EDITOR'; otherwise the command will
accept data until EOF or, unless the notes field is being edited, the first new line.
Please note that when editing interactively, the contents of the field may be
saved on disk in tmp files or in editor swap files, depending on your system
configuration.

The 'generate' subcommand will create a randomly generated password for the
chosen key name, and optionally add a url and username while inserting the
generated password.

The 'rm' command will remove the specified entry, and the 'duplicate' command
will create a duplicate entry of the one specified, but with a different 'ID'.

Backup
~~~~~~
The 'export' subcommand will dump all account information including
passwords to stdout (unencrypted) in CSV format.  The optional
'--fields=FIELDLIST' argument may contain a comma-separated subset of the
following fields:

  id, url, username, password, extra, name, fav, id, grouping, group,
  fullname, last_touch, last_modified_gmt, attachpresent

The 'import' subcommand does the reverse: accounts from an unencrypted
CSV file are uploaded to the server.

It is recommended that such backups be encrypted at rest, for example by
piping to and from gpg.

Shared Folder Commands
~~~~~~~~~~~~~~~~~~~~~~
The 'share' command and its accompanying subcommands can be used to manipulate
shared folders, if available to the (enterprise or premium) user.  The
'userls', 'useradd', 'usermod', and 'userdel' subcommands may be used to
query and modify membership of the shared folder, while the 'create' and 'rm'
share subcommands may be used to add new, or delete existing shared folders.
The normal 'generate' and 'edit' commands may be used to edit accounts within
the shared folder.

The 'share limit' command may be used to manipulate account access lists on the
share for a specific user.  Running with no arguments will display the
current access levels for a user.  The '--add', '--rm', and '--clear' options
may be used to add to, remove from, or reset the list.  Passing '--allow' or
'--deny' will make the list a whitelist or blacklist, respectively.

Clipboard
~~~~~~~~~
Commands that take a '-c' or '--clip' option will copy the output to the
clipboard, using *xclip*(1) or *xsel*(1) on X11-based systems, *pbcopy*(1)
on OSX, or *putclip* on Cygwin. The command to be used can be overridden by
specifying the `LPASS_CLIPBOARD_COMMAND` environment variable.

Color Output
~~~~~~~~~~~~
The '--color' option controls colored output to the terminal.  By default,
commands will use '--color=auto', in which color output is used unless the
output is not a tty (for example, when passed to a pipe or file).  If 'always'
is used, colors are produced regardless of the output detection.  If 'never'
is used, no color escape sequences are emitted.

Configuration
~~~~~~~~~~~~~
'lpass' stores configuration in the following locations, in descending
order of precedence:

* The directory '$LPASS_HOME', if set
* '$XDG_CONFIG_HOME/lpass', '$XDG_DATA_HOME/lpass', and
  '$XDG_RUNTIME_DIR/lpass' (or equivalent defaults), if at least
  '$XDG_RUNTIME_DIR' is set
* '$HOME/.lpass'

All configuration may be specified via environment variables. Alternatively, a
set of environment variable overrides may be specified in '$LPASS_HOME/env' in
the form of:

[verse]
VARIABLE1=VALUE1
VARIABLE2=VALUE2
...

Aliases
~~~~~~~
Default options can be specified for any command by creating command aliases.
Command aliases are stored in the configuration directory (see previous
section) with files named 'alias.*command*'.  For example, to force the
'ls' subcommand to never use color, make an 'alias.ls' file with the
appropriate option:
----
echo 'ls --color=never' > ~/.config/lpass/alias.ls
----

Similarly, new subcommands can be created based on built-in subcommands with
options.  Thus,
----
echo 'show --password -c' > ~/.config/lpass/alias.passclip
----

would create a 'passclip' subcommand that copies your password onto the
clipboard.

ENVIRONMENT VARIABLES
---------------------
The following environment variables may be used for configuration as described
in the section above:

* 'LPASS_HOME'
* 'LPASS_AUTO_SYNC_TIME'
* 'LPASS_AGENT_TIMEOUT'
* 'LPASS_AGENT_DISABLE'
* 'LPASS_PINENTRY'
* 'LPASS_DISABLE_PINENTRY'
* 'LPASS_ASKPASS'
* 'LPASS_CLIPBOARD_COMMAND'

EXAMPLES
--------

In the following examples, "$" indicates a shell prompt while "#"
indicates a comment.

----
# login, generate and retrieve a new password
$ lpass login user@example.com
$ lpass generate work/email 20
G</Czo4*c;A+g;KIHkO-
$ lpass ls work
work
    email [id: 140613939481239829]
$ lpass show -p email
G</Czo4*c;A+g;KIHkO-

# Copy password to clipboard
$ lpass show -cp email

# Add a new secure note
$ lpass add --note-type=ssn info/my-ssn
# (editor shows a template file; fill in and save)
Name: info/my-ssn
NoteType: Social Security
Name: My Name Goes Here
Number: 123-45-6789
Notes:    # Add notes below this line
Any notes I want to save go here.
~
~

$ lpass show my-ssn
info/my-ssn [id: 0]
Number: 123-45-6789
Name: My Name Goes Here
NoteType: Social Security
Notes:
Any notes I want to save go here.

# Display a secure note attachment
$ lpass show my-secure-note
info/my-secure-note [id: 1426405543365295118]
URL: http://sn
att-1426405543365295118-94690: travel-flight.pdf
$ lpass show my-secure-note --attach att-1426405543365295118-94690
"travel-flight.pdf" is a binary file, print it anyway (or save)?  [y/n/S] s
Wrote 122864 bytes to "travel-flight.pdf"

# Display secure note attachment to standard output
$ lpass show my-secure-note --attach att-1426405543365295118-94690 --quiet
[... binary data on stdout ...]

# Add an account non-interactively by creating the proper template
printf "Username: wizard97\nPassword: vJwhFfBBtn8hj4" | \
    lpass add Facebook --non-interactive
----




CHANGELOG.md
# Version 1.6.1
* Fixed a bug that prevented groups from being added to shared folders whenever any of the users did not have a public key

# Version 1.6.0
* Improve sharing key generation
* Remove SHA-256 support for login hashes
* Fix password change failure
* Replace logic of reusing key as IV
* Fix incorrect pointer-to-pointer argument passing that would yield compilation errors on GCC-14

# Version 1.5.0
* Add support for URL logging
* Fix segmentation faults on Mac OS X
* Fix synchronization issues, which caused empty IDs
* Fix feature flags file deletion on logout

# Version 1.4.0
* Add support for writing encrypted URLs
* Add support for feature flags
* Fix shared folder username

# Version 1.3.7
* Add support for reading encrypted URLs (Tibor Komlossy)
* Fix GCC 10 compatibility issue #532 (Tibor Komlossy)

# Version 1.3.6
* Fix version (Béla Ormos)

# Version 1.3.5
* Updating certificate hashes (Béla Ormos)

# Version 1.3.4
* Updating post parameter (Gergely Der)

# Version 1.3.3
 * Decrease the time for the cli app to do things (Wesley Schwengle)
 * 'blob_load' refactor (Wesley Schwengle)
 * Fixed bug where logout requires login (Wesley Schwengle)
 * Fix non-default PKG_CONFIG_PATH on macOS (Alyssa Ross)

# Version 1.3.2
 * Don't require using make (Eli Schwartz)
 * Disable IPv6 support (Wesley Schwengle)
 * Link against Brew Curl on MacOS (Tom Sullivan)
 * Autogenerate versions from git (Wesley Schwengle/Eli Schwartz)
 * Remove memory leak in `config_path_for_type` (Tom Sullivan)
 * Install bash-completions in PREFIX dir (Wesley Schwengle)
 * Include `libgen.h` for BSD builds (Tom Sullivan)
 * Create subdirectories when determining config path (Tom Sullivan)
 * Only show basename in usage (William Casarin)
 * Fix segmentation fault on BSD while running `make test` (Björn Ketelaars)
 * README.md updates:
   * Brew formula installation (Thomas Haggett)
   * Ubuntu Xenial dependencies (Nick Timkovich)
   * Debian stable/testing and Ubuntu dependencies (Wesley Schwengle)

# Version 1.3.1
 * Revert "pins: remove GlobalSign R1/R3 pins" from Robert Copeland
 * Readme update from Wesley Schwengle
 * Add Dockerfile to create a clean build environment from Wesley Schwengle
 * Missing dependencies in readme
 * Added CLion project files to ignore list

# Version 1.3.0
 * `lpass show` now supports `--json` format
 * `lpass show` now supports `--quiet` flag to suppress prompts,
   from Pau Sanchez
 * `lpass import` has `--keep-dupes` flag which will preserve duplicate
   accounts on import
 * `LPASS_PINENTRY` environment variable may now be used to set custom
   path to pinentry, from Martynas Mickevičius
 * Build fix for aarch64 and others from Natanael Copa
 * New fish completions from Israel Chauca Fuentes
 * Zsh completions from Richard Hillmann
 * Brew build instructions updates from Roger D. Winans
 * bugfix: site notes now show up in notes textarea instead of fields
 * spelling fixes from Josh Soref

# Version 1.2.2
 * `lpass ls --format` now supports "%al" to show URL, from Yikai Zhao
 * $VISUAL can be used in preference to $EDITOR, from Wesley Schwengle
 * `lpass edit` can now directly edit multiline ssh keys
 * fields are now preserved when edited with `lpass edit`
 * Bugfix: use-after-free in http.c fixed, from Björn Ketelaars
 * Bugfix: command-line completion now works for names with spaces
 * Bugfix: loading attachments from shared folders now works, from Spencer
   Whyte
 * Debian packing updates from Hannes Hörl
 * Documentation updates from Darragh Grealish and Steven Liekens

# Version 1.2.1
 * Bugfix: fix regression with ```lpass show``` not displaying all fields
   for secure notes
 * Use sysctl instead of procfs for pid-to-cmd on some versions of BSD,
   from Thomas Hurst
 * Build: fix build for test binaries on OpenBSD, from Björn Ketelaars

# Version 1.2.0
 * ```lpass show``` now supports new-style multiline ssh keys
 * ```lpass export``` now supports --fields=FIELDLIST argument to
   control output, with patches from Kyle Burton
 * ```lpass ls``` now always shows empty shared folders
 * ```lpass edit``` can now set the 'master password reprompt' field in sites
 * ```lpass share create``` now shows the created share name
 * Bugfix: crash in `lpass show` fixed by Kyle Burton
 * build fixes for termux and documentation updates, from Christian Rondeau
 * documentation updates for Ubuntu from Craig Menning and Glenn Oppegard
 * Test suite now included covering basic operations

# Version 1.1.2
 * Bugfix: crash with ```lpass logout --color=never``` fixed
 * Bugfix: ```lpass add``` with secure notes works again
 * Bugfix: sort order in ```lpass ls``` is now consistent whether or not
   colors are used
 * Documentation has been enhanced to describe aliases and more options,
   with patches by Eric B. Hymowitz.
 * Build: debian package fixed for rebuild issues and missing build
   dependencies

# Version 1.1.1
 * Bugfix: fix crash in ```lpass show``` for secure notes without attachments
 * Build: fix build on OpenBSD
 * Build: fix build when using LibreSSL

# Version 1.1.0
 * New command ```lpass import``` can import an existing csv file (or output
   from ```lpass export``` into the vault
 * ```lpass show``` and ```lpass ls``` learned a ```--format``` argument
   to enable user-specified printf-style formats
 * Bash completions will now complete field names if ```--field``` is
   specified after the account name
 * Build: cmake now used for building, by Filippo Cucchetto and
   with fixes by Eli Schwartz
 * Build: lpass has been updated to work with OpenSSL 1.1; please note
   that libcurl-openssl must also be linked against the same version
   in order to avoid mysterious segfaults
 * Bugfix: crash in ```lpass ls -l``` with no last_modified_gmt fixed
 * Bugfix: secure notes editing with "Name" fields now works properly
 * Bugfix: editing secure note names now works (github #106)
 * Bugfix: lpass-created server secure notes are now compatible with the plugin

lpass.c
/*
 * lpass - lastpass command line utility
 *
 * Copyright (C) 2014-2018 LastPass.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * In addition, as a special exception, the copyright holders give
 * permission to link the code of portions of this program with the
 * OpenSSL library under certain conditions as described in each
 * individual source file, and distribute linked combinations
 * including the two.
 *
 * You must obey the GNU General Public License in all respects
 * for all of the code used other than OpenSSL.  If you modify
 * file(s) with this exception, you may extend this exception to your
 * version of the file(s), but you are not obligated to do so.  If you
 * do not wish to do so, delete this exception statement from your
 * version.  If you delete this exception statement from all source
 * files in the program, then also delete it here.
 *
 * See LICENSE.OpenSSL for more details regarding this exception.
 */
#include "process.h"
#include "cmd.h"
#include "string.h"
#include "util.h"
#include "http.h"
#include "config.h"
#include "terminal.h"
#include "version.h"
#include "log.h"
#include <sys/stat.h>
#include <getopt.h>
#include <unistd.h>

#if (defined(__APPLE__) && defined(__MACH__)) || defined(__FreeBSD__) || defined(__DragonFly__) || defined(__NetBSD__) || defined(__OpenBSD__)
#include <libgen.h>
#endif

#define CMD(name) { #name, cmd_##name##_usage, cmd_##name }
static struct {
	const char *name;
	const char *usage;
	int (*cmd)(int, char**);
} commands[] = {
	CMD(login),
	CMD(logout),
	CMD(passwd),
	CMD(show),
    CMD(history),
	CMD(ls),
	CMD(mv),
	CMD(add),
	CMD(edit),
	CMD(generate),
	CMD(duplicate),
	CMD(rm),
	CMD(status),
	CMD(sync),
	CMD(export),
	CMD(import),
	CMD(share)
};
#undef CMD

static void version(void)
{
	terminal_printf("LastPass CLI v" LASTPASS_CLI_VERSION "\n");
}

static void help(void)
{
	terminal_printf("Usage:\n");
	printf("  %s {--help|--version}\n", ARGV[0]);
	for (size_t i = 0; i < ARRAY_SIZE(commands); ++i)
		printf("  %s %s\n", basename(ARGV[0]), commands[i].usage);
}

static int global_options(int argc, char *argv[])
{
	static struct option long_options[] = {
		{"version", no_argument, NULL, 'v'},
		{"help", no_argument, NULL, 'h'},
		{0, 0, 0, 0}
	};
	int option;
	int option_index;

	while ((option = getopt_long(argc, argv, "vh", long_options, &option_index)) != -1) {
		switch (option) {
			case 'v':
				version();
				return 0;
			case 'h':
				version();
				printf("\n");
				help();
				return 0;
			case '?':
				help();
				return option == 'h';
		}
	}

	help();
	return 1;
}

static void expand_aliases(int *argc, char ***argv)
{
	int i;
	const char *alias = (*argv)[0];
	char **new_argv = NULL;
	int argv_alloced;
	int new_argc = 0;
	_cleanup_free_ char *config_name;

	xasprintf(&config_name, "alias.%s", alias);

	_cleanup_free_ char *alias_val = config_read_string(config_name);
	if (!alias_val)
		return;

	trim(alias_val);

	/* split commandline and prepend to argv */
	argv_alloced = 0;
	new_argv = xcalloc(*argc + 1, sizeof(*new_argv));

	char *tok = strtok(alias_val, " \t");
	while (tok) {
		if (new_argc >= argv_alloced) {
			argv_alloced += 16;
			new_argv = xreallocarray(new_argv,
				argv_alloced + *argc + 1, sizeof(*new_argv));
		}
		new_argv[new_argc++] = xstrdup(tok);
		tok = strtok(NULL, " \t");
	}

	/* copy in remaining items from argc */
	for (i=1; i < *argc; i++) {
		new_argv[new_argc++] = xstrdup((*argv)[i]);
	}
	new_argv[new_argc] = 0;
	*argv = new_argv;
	*argc = new_argc;
}

static int process_command(int argc, char *argv[])
{
	expand_aliases(&argc, &argv);

	for (size_t i = 0; i < ARRAY_SIZE(commands); ++i) {
		if (argc && !strcmp(argv[0], commands[i].name))
			return commands[i].cmd(argc, argv);
	}
	help();
	return 1;
}

static void load_saved_environment(void)
{
	_cleanup_free_ char *env = NULL;

	env = config_read_string("env");
	if (!env)
		return;

	for (char *tok = strtok(env, "\n"); tok; tok = strtok(NULL, "\n")) {
		char *equals = strchr(tok, '=');
		if (!equals || !*equals) {
			warn("The environment line '%s' is invalid.", tok);
			continue;
		}
		*equals = '\0';
		if (setenv(tok, equals + 1, true))
			warn_errno("The environment line '%s' is invalid.", tok);
	}
}

int main(int argc, char *argv[])
{
	/* For process.h to function. */
	ARGC = argc;
	ARGV = argv;

	/* Do not remove this umask. Always keep at top. */
	umask(0077);

	if (http_init())
		die("Unable to initialize curl");

	load_saved_environment();

	if (argc >= 2 && argv[1][0] != '-')
		return process_command(argc - 1, argv + 1);

	return global_options(argc, argv);
}
